src/general/ShuffleBag.cs

```diff
@@ -0,0 +1,137 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+
+public class ShuffleBag<T> : IEnumerable<T>
+{
+    private Random random;
+    private List<T> initialContent;
+    private List<T> currentContent;
+
+    public ShuffleBag(Random random)
+    {
+        initialContent = new List<T>();
+        currentContent = new List<T>();
+        this.random = random;
+    }
+
+    public void Clear()
+    {
+        initialContent.Clear();
+        currentContent.Clear();
+    }
+
+    public void Add(T element)
+    {
+        initialContent.Add(element);
+        currentContent.Add(element);
+    }
+
+    public bool Remove(T element)
+    {
+        currentContent.Remove(element);
+        return initialContent.Remove(element);
+    }
+
+    public void RemoveAll(Predicate<T> f)
+    {
+        initialContent.RemoveAll(f);
+    }
+
+    public void FillAndShuffle()
+    {
+        currentContent.Clear();
+
+        foreach (var element in initialContent)
+            currentContent.Add(element);
+
+        Shuffle();
+    }
+
+    public T PickAndDrop()
+    {
+        var drawnElement = Pick();
+        Drop();
+
+        return drawnElement;
+    }
+
+    public IEnumerator<T> GetEnumerator()
+    {
+        return new ShuffleBagEnumerator<T>(this);
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return GetEnumerator();
+    }
+
+    /// <summary>
+    ///   Fisher–Yates Shuffle Alg. (https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle)
+    ///   Provides a uniform random shuffle in O(n)
+    /// </summary>
+    private void Shuffle()
+    {
+        for (int i = 0; i < currentContent.Count - 2; i++)
+        {
+            int j = random.Next(i, currentContent.Count);
+
+            T swapedElement = currentContent[i];
+            currentContent[i] = currentContent[j];
+            currentContent[j] = swapedElement;
+        }
+    }
+
+    private T Pick()
+    {
+        if (currentContent.Count == 0)
+        {
+            FillAndShuffle();
+        }
+
+        return currentContent[currentContent.Count - 1];
+    }
+
+    private bool Drop()
+    {
+        if (currentContent.Count == 0)
+            throw new InvalidOperationException("Cannot drop from empty bag!");
+
+        currentContent.RemoveAt(currentContent.Count - 1);
+        return currentContent.Count == 0;
+    }
+
+    private class ShuffleBagEnumerator<T1> : IEnumerator<T1>
+    {
+        private ShuffleBag<T1> sourceBag;
+
+        public ShuffleBagEnumerator(ShuffleBag<T1> sourceBag)
+        {
+            this.sourceBag = sourceBag;
+        }
+
+        public T1 Current => sourceBag.Pick();
+
+        object IEnumerator.Current => Current;
+
+        public void Dispose()
+        {
+            // If we have exhausted the bag, fill and shuffle it for the next one.
+            if (sourceBag.currentContent.Count == 0)
+                sourceBag.FillAndShuffle();
+
+            // TODO CLEAR MEMORY
```

Comment:
what is meant by clear memory?

Reply:
Given that I'm not familiar with enumerators (especially in C#), there may be some memory cleaning to do when disposing the enumerator. If not, then nothing is to be done, but I added this comment just in case.

Made it a clearer comment in the method documentation though.

Reply:
I mean this class doesn't hold any disposable resources, so there is nothing to dispose here.
One thing I do want to criticise here is that this relies on being *disposed* to trigger bag re-fill, which I fear is actually problematic *if* someone gets the enumerator without using a foreach as I think that automatically will call dispose (I *think*).

Reply:
You can actually trigger bag refill with the `Reset()` method. Dispose does so as well because foreach does not call `Reset` (writing these lines I realize I should perhaps call `this.Reset()` instead, btw?).

After that, if the coder wants to keep a non-reset enumerator... well, that's possible, as for a list enumerator you do not reset.

Reply:
I think it would make a ton more sense to reset the sequence when *asking* for an enumerator, instead of this way. C# is not C++ so RAII is not really a thing, ie. object destruction needs to not be an important operation / relied upon happening...

Reply:
That's a fair point. I made the filling occur at the creation of the enumerator if the bag is empty.